### PR TITLE
Features/integrate and merge pypsa eur

### PR DESCRIPTION
**Pull request by ClaraBuettner**
This is a cleand version of PR #122. The other PR also included commits from other branches, some of them are CI-branches and will never be merged into dev. 
To avoid problems in the future, I created this cleaned version that only includes the commits necessary to integrate pypsaeur. 
Before this PR is merged, I will check again if it includes all needed changes (using git diff /old/branch).

### CODE REVIEW

**Comment by ClaraBuettner** [comment]
@CarlosEpia: I found a solution to get the files that were not accessible within pypsaeur. 
The complex parts of pypsaeur were already running successfully, I am just waiting until the solver solved the optimization problem.

File: src/egon/data/airflow/dags/pipeline.py [hunk 1]
```diff
@@ -352,30 +355,47 @@
...
+    geothermal_potential_germany = GeothermalPotentialGermany(
+        dependencies=[
+            data_bundle,
+            district_heating_areas,
+        ]
+        )
+
+    # Deal with electrical neighbours
+    foreign_lines = ElectricalNeighbours(
+        dependencies=[prepare_pypsa_eur, tyndp_data]
```
Comment (CarlosEpia) [inline src/egon/data/airflow/dags/pipeline.py]: I think ElectricalNeighbours requires that the task _electrical_neighbours_egon100_ is previously executed. The mentioned task belongs to _run_pypsaeur_ but the dependency points to _ PreparePypsaEur_.

Comment (ClaraBuettner) [reply src/egon/data/airflow/dags/pipeline.py]: There is a dependency the other way around: electrical_neighbors_egon100 depends on ElectricalNeighbours. All AC buses are now created and inserted within ElectricalNeighbours. In electrical_neighbors_egon100, components are attached to these buses, considering the indices inserted by ElectricalNeighbours. 
Could you explain why you see the need for the dependency you mentioned?

**Comment by CarlosEpia** [comment]
As discussed, there are problems to create the virtual environment where pypsa-eur is executed. Altermativaly, to test other tasks I used the results from the instance "powerd-run-de", but the task pypsaeur.electrical-neighbours-egon100 is failing with this error message:
[2024-08-05, 09:59:20 UTC] {taskinstance.py:2905} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.SyntaxError: unterminated quoted identifier at or near ""
        "
LINE 5: ... bus_id NOT IN (SELECT bus_i FROM osmtgmod_results.bus_***"
                                                                      ^
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 465, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 432, in _execute_callable
    return execute_callable(context=context, **execute_callable_kwargs)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/models/baseoperator.py", line 401, in wrapper
    return func(self, *args, **kwargs)
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/datasets/__init__.py", line 216, in skip_task
    result = super(type(task), task).execute(*xs, **ks)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/models/baseoperator.py", line 401, in wrapper
    return func(self, *args, **kwargs)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/operators/python.py", line 235, in execute
    return_value = self.execute_callable()
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/operators/python.py", line 252, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/datasets/pypsaeur/__init__.py", line 302, in electrical_neighbours_egon100
    neighbor_reduction()
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/datasets/pypsaeur/__init__.py", line 494, in neighbor_reduction
    foreign_ac_buses = db.select_dataframe(
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/db.py", line 188, in select_dataframe
    df = pd.read_sql(sql, engine(), index_col=index_col)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/pandas/io/sql.py", line 663, in read_sql
    return pandas_sql.read_query(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/pandas/io/sql.py", line 1738, in read_query
    result = self.execute(sql, params)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/pandas/io/sql.py", line 1562, in execute
    return self.con.exec_driver_sql(sql, *args)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1770, in exec_driver_sql
    return self._exec_driver_sql(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1674, in _exec_driver_sql
    ret = self._execute_context(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) unterminated quoted identifier at or near ""
        "
LINE 5: ... bus_id NOT IN (SELECT bus_i FROM osmtgmod_results.bus_***"
                                                                      ^
[SQL: 
        SELECT * FROM grid.egon_etrago_bus
        WHERE carrier = 'AC' AND v_nom = 380
        AND country!= 'DE' AND scn_name ='eGon100RE'
        AND bus_id NOT IN (SELECT bus_i FROM osmtgmod_results.bus_data"
        ]
(Background on this error at: https://sqlalche.me/e/14/f405)
[2024-08-05, 09:59:20 UTC] {taskinstance.py:1206} INFO - Marking task as FAILED. dag_id=egon-***-processing-pipeline, task_id=pypsaeur.electrical-neighbours-egon100, run_id=manual__2024-08-01T18:11:40.363517+00:00, execution_date=20240801T181140, start_date=20240805T095907, end_date=20240805T095920
[2024-08-05, 09:59:20 UTC] {standard_task_runner.py:110} ERROR - Failed to execute job 2543 for task pypsaeur.electrical-neighbours-egon100 ((psycopg2.errors.SyntaxError) unterminated quoted identifier at or near ""
        "
LINE 5: ... bus_id NOT IN (SELECT bus_i FROM osmtgmod_results.bus_***"
                                                                      ^
[SQL: 
        SELECT * FROM grid.egon_etrago_bus
        WHERE carrier = 'AC' AND v_nom = 380
        AND country!= 'DE' AND scn_name ='eGon100RE'
        AND bus_id NOT IN (SELECT bus_i FROM osmtgmod_results.bus_data"
        ]
(Background on this error at: https://sqlalche.me/e/14/f405); 2641694)
[2024-08-05, 09:59:20 UTC] {local_task_job_runner.py:240} INFO - Task exited with return code 1

Do you know what could be happening?

**Comment by ClaraBuettner** [comment]
A bracket was missing, I fixed that in the latest commit.
Did you already find out any reason why the creation of the environment might have failed?

**Comment by ClaraBuettner** [comment]
@CarlosEpia: I found out why the environment could not be created. It was fixed in https://github.com/openego/powerd-data/pull/293/commits/a8f435141da77c8983a1e1ec9225e5c01d4d98e0

I also added the checks if pypsaeur should be executed or not. So download, prepare and solve network are now only executed if the cli parameter is set to True.

**Comment by CarlosEpia** [comment]
> @CarlosEpia: I found out why the environment could not be created. It was fixed in [a8f4351](https://github.com/openego/powerd-data/commit/a8f435141da77c8983a1e1ec9225e5c01d4d98e0)
> 
> I also added the checks if pypsaeur should be executed or not. So download, prepare and solve network are now only executed if the cli parameter is set to True.

Great. I will give it a try now.

**Comment by CarlosEpia** [comment]
It looks like we are missing a dependency. Electrical neighbours failed with this error message:

  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/datasets/electrical_neighbours.py", line 832, in grid
    central_buses = buses(scenario, sources, targets)
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/datasets/electrical_neighbours.py", line 161, in buses
    db.execute_sql(sql_delete)
  File "/home/powerd/egon100re-run/powerd-data/src/egon/data/db.py", line 73, in execute_sql
    con.execute(text(sql_string))
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1385, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) relation "osmtgmod_results.bus_***" does not exist
LINE 8:             FROM  osmtgmod_results.
                          ^
[SQL: 
        DELETE FROM grid.
            egon_etrago_bus
        WHERE country != 'DE' AND scn_name = 'eGon100RE'
        AND carrier = 'AC'
        AND bus_id NOT IN (
            SELECT bus_i
            FROM  osmtgmod_results.
            bus_data)
        ]
(Background on this error at: https://sqlalche.me/e/14/f405)
[2024-08-08, 08:36:28 UTC] {taskinstance.py:1206} INFO - Marking task as FAILED. dag_id=egon-***-processing-pipeline, task_id=electrical_neighbours.grid, run_id=manual__2024-08-08T08:22:11.503923+00:00, execution_date=20240808T082211, start_date=20240808T083628, end_date=20240808T083628
[2024-08-08, 08:36:28 UTC] {standard_task_runner.py:110} ERROR - Failed to execute job 2590 for task electrical_neighbours.grid ((psycopg2.errors.UndefinedTable) relation "osmtgmod_results.bus_***" does not exist
LINE 8:             FROM  osmtgmod_results.
                          ^
[SQL: 
        DELETE FROM grid.
            egon_etrago_bus
        WHERE country != 'DE' AND scn_name = 'eGon100RE'
        AND carrier = 'AC'
        AND bus_id NOT IN (
            SELECT bus_i
            FROM  osmtgmod_results.
            bus_data)
        ]
(Background on this error at: https://sqlalche.me/e/14/f405); 108608)
[2024-08-08, 08:36:28 UTC] {local_task_job_runner.py:240} INFO - Task exited with return code 1

**Comment by ClaraBuettner** [comment]
The table that seems to be missing is created in osmtgmod. That is executed very early in the process. Is there a dependency to that?

**Comment by CarlosEpia** [comment]
> The table that seems to be missing is created in osmtgmod. That is executed very early in the process. Is there a dependency to that?

 Electrical neighbours depends only on
- tyndp.download which have no dependency
- and pypsaeur.prepare-network which depends only on download tasks (era5.insert-weather-cells and data_bundle.download)

**Comment by ClaraBuettner** [comment]
> > The table that seems to be missing is created in osmtgmod. That is executed very early in the process. Is there a dependency to that?
> 
> Electrical neighbours depends only on
> 
>     * tyndp.download which have no dependency
> 
>     * and pypsaeur.prepare-network which depends only on download tasks (era5.insert-weather-cells and data_bundle.download)

I added a dependency to osmtgmod

**Comment by CarlosEpia** [comment]
The task pypsa-eur.prepare_network is failing with this error message:
[2024-08-12, 14:34:41 UTC] {taskinstance.py:1206} INFO - Marking task as FAILED. dag_id=egon-***-processing-pipeline, task_id=pypsaeur.prepare-network, run_id=manual__2024-08-12T13:47:31.104397+00:00, execution_date=20240812T134731, start_date=20240812T142911, end_date=20240812T143441
[2024-08-12, 14:34:41 UTC] {standard_task_runner.py:110} ERROR - Failed to execute job 2780 for task pypsaeur.prepare-network (
| Command '['snakemake', '-j1', '--directory', PosixPath('run-pypsa-
| eur'), '--snakefile', PosixPath('run-pypsa-eur/Snakefile'), '--use-
| conda', '--conda-frontend=conda', 'prepare']' returned non-zero exit
| status 1.
| Output:
| Errors:
|   Building DAG of jobs...
|   Creating conda environment pypsa-eur/envs/environment.yaml...
|   Downloading and installing remote packages.
|   CreateCondaEnvironmentException:
|   Could not create conda environment from /home/powerd/egon100re-
|   run/run-pypsa-eur/pypsa-eur/envs/environment.yaml:
|   Retrieving notices: ...working... done
|   Channels:
|    - conda-forge
|    - bioconda
|    - defaults
|   Platform: linux-64
|   Collecting package metadata (repodata.json): ...working... done
|   Solving environment: ...working... done
|   Preparing transaction: ...working... done
|   Verifying transaction: ...working... done
|   Executing transaction: ...working...
|   done
|   Installing pip dependencies: ...working... Pip subprocess error:
|   ERROR: Ignored the following yanked versions: 11.0.0b1
|   ERROR: Could not find a version that satisfies the requirement
|   gurobipy==10.0.0 (from versions: 11.0.0, 11.0.1, 11.0.2, 11.0.3)
|   ERROR: No matching distribution found for gurobipy==10.0.0
|   failed
|   CondaEnvException: Pip failed; 3956971)
[2024-08-12, 14:34:41 UTC] {local_task_job_runner.py:240} INFO - Task exited with return code 1
[2024-08-12, 14:34:41 UTC] {logging_mixin.py:188} WARNING - /home/powerd/egon100re-run/venv2/lib/python3.8/site-packages/airflow/models/baseoperator.py:1297 AirflowProviderDeprecationWarning: Call to deprecated class PostgresOperator. (Please use `airflow.providers.common.sql.operators.sql.SQLExecuteQueryOperator`.Also, you can provide `hook_params={'schema': <***base>}`.)
[2024-08-12, 14:34:41 UTC] {taskinstance.py:3503} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2024-08-12, 14:34:41 UTC] {local_task_job_runner.py:222} ▲▲▲ Log group end

@ClaraBuettner do you know why it is not finding the gurobipy==10.0.0?? I already checked and that exact version exist in pip.

**Comment by ClaraBuettner** [comment]
No, I don't know why this happened. I use the same version and was also able to install it again. I can have a look at it.  
On the server, the task looks like it was successful, but there is still the error message. Did you mark it as successful?

**Comment by ClaraBuettner** [comment]
It looks like the default python version from conda changed to python3.12 which is not compatible to gurobipy <0.11. 
I am still wondering how this could happen, but I guess it is quite easy to fix

**Comment by CarlosEpia** [comment]
I gave it another try after deleting the folder _run-pypsa-eur_ and after almost 4 hours the process prepare-network failed with this error message:

|       reason: Missing output files:
|   resources/gas_input_locations_s_37_simplified.csv,
|   resources/gas_input_locations_s_37.geojson; Input files updated by
|   another job: data/gas_network/scigrid-
|   gas/data/IGGIELGN_BorderPoints.geojson,
|   resources/regions_offshore_elec_s_37.geojson,
|   resources/regions_onshore_elec_s_37.geojson,
|   data/gas_network/scigrid-gas/data/IGGIELGN_Storages.geojson
|       wildcards: simpl=, clusters=37
|       resources: tmpdir=/tmp, mem_mb=2000, mem_mib=1908
|   Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Downloading from remote: globalenergymonitor.org/wp-
|   content/uploads/2023/07/Europe-Gas-Tracker-2023-03-v3.xlsx
|   ProtocolError: ('Connection broken: IncompleteRead(457215 bytes
|   read, 575062 more expected)', IncompleteRead(457215 bytes read,
|   575062 more expected)), attempt 1/3 failed - retrying in 3
|   seconds...
|   ProtocolError: ('Connection broken: IncompleteRead(408063 bytes
|   read, 624214 more expected)', IncompleteRead(408063 bytes read,
|   624214 more expected)), attempt 2/3 failed - retrying in 6
|   seconds...
|   ProtocolError: ('Connection broken: IncompleteRead(408064 bytes
|   read, 624213 more expected)', IncompleteRead(408064 bytes read,
|   624213 more expected)), attempt 3/3 failed - giving up!
|   WorkflowError:
|   ProtocolError: ('Connection broken: IncompleteRead(408064 bytes
|   read, 624213 more expected)', IncompleteRead(408064 bytes read,
|   624213 more expected))
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/decorator.py", line 232, in fun
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 161, in retry_decorator
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 218, in retry_call
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 31, in __retry_internal
|   [Tue Aug 13 23:58:55 2024]
|   Error in rule prepare:
|       jobid: 0
|       conda-env: /home/powerd/egon100re-run/run-pypsa-
|   eur/.snakemake/conda/fd40379c484bc62eb71ff713b0d3f10a
|       shell:
|           snakemake --directory=pypsa-eur/ --snakefile=pypsa-
|   eur/Snakefile -call prepare_sector_networks
|           (one of the commands exited with non-zero exit code; note
|   that snakemake uses bash strict mode!)
|   Shutting down, this might take some time.
|   Exiting because a job execution failed. Look above for error message
|   Complete log: /home/powerd/egon100re-run/run-pypsa-
|   eur/.snakemake/log/2024-08-13T200147.595763.snakemake.log; 37451)

**Comment by CarlosEpia** [comment]
The task pypsaeur-prepare-network failed with the next error message:
|   sector/nuts/NUTS_RG_10M_2013_4326_LEVL_2.geojson
|       wildcards: simpl=, clusters=37, planning_horizons=2050
|       resources: tmpdir=/tmp, mem_mb=1000, mem_mib=954
|   Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd40379
|   c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/distributed/node.py:182: UserWarning: Port 8787 is already
|   in use.
|   Perhaps you already have a cluster running?
|   Hosting the HTTP server on port 46867 instead
|     warnings.warn(
|   /home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd40379c48
|   4bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/distributed/node.py:182: UserWarning: Port 8787 is already
|   in use.
|   Perhaps you already have a cluster running?
|   Hosting the HTTP server on port 41917 instead
|     warnings.warn(
|   Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS" (6)Error.  nthreads
|   cannot be larger than environment variable "NUMEXPR_MAX_THREADS"
|   (6)Error.  nthreads cannot be larger than environment variable
|   "NUMEXPR_MAX_THREADS" (6)Error.  nthreads cannot be larger than
|   environment variable "NUMEXPR_MAX_THREADS"
|   (6)/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd40379
|   c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/geopandas/geodataframe.py:1528: SettingWithCopyWarning:
|   A value is trying to be set on a copy of a slice from a DataFrame.
|   Try using .loc[row_indexer,col_indexer] = value instead
|   See the caveats in the documentation:
|   https://pandas.pydata.org/pandas-
|   docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
|     super().__setitem__(key, value)
|   /home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd40379c48
|   4bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/geopandas/geodataframe.py:1528: SettingWithCopyWarning:
|   A value is trying to be set on a copy of a slice from a DataFrame.
|   Try using .loc[row_indexer,col_indexer] = value instead
|   See the caveats in the documentation:
|   https://pandas.pydata.org/pandas-
|   docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
|     super().__setitem__(key, value)
|   Downloading from remote:
|   zenodo.org/records/10356004/files/ENSPRESO_BIOMASS.xlsx
|   ProtocolError: ('Connection broken: IncompleteRead(648280 bytes
|   read, 15114049 more expected)', IncompleteRead(648280 bytes read,
|   15114049 more expected)), attempt 1/3 failed - retrying in 3
|   seconds...
|   ProtocolError: ('Connection broken: IncompleteRead(641916 bytes
|   read, 15120413 more expected)', IncompleteRead(641916 bytes read,
|   15120413 more expected)), attempt 2/3 failed - retrying in 6
|   seconds...
|   ProtocolError: ('Connection broken: IncompleteRead(621428 bytes
|   read, 15140901 more expected)', IncompleteRead(621428 bytes read,
|   15140901 more expected)), attempt 3/3 failed - giving up!
|   WorkflowError:
|   ProtocolError: ('Connection broken: IncompleteRead(621428 bytes
|   read, 15140901 more expected)', IncompleteRead(621428 bytes read,
|   15140901 more expected))
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/decorator.py", line 232, in fun
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 161, in retry_decorator
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 218, in retry_call
|     File "/home/powerd/egon100re-run/run-pypsa-eur/.snakemake/conda/fd
|   40379c484bc62eb71ff713b0d3f10a/lib/python3.11/site-
|   packages/reretry/api.py", line 31, in __retry_internal
|   [Wed Aug 14 21:15:32 2024]
|   Error in rule prepare:
|       jobid: 0
|       conda-env: /home/powerd/egon100re-run/run-pypsa-
|   eur/.snakemake/conda/fd40379c484bc62eb71ff713b0d3f10a
|       shell:
|           snakemake --directory=pypsa-eur/ --snakefile=pypsa-
|   eur/Snakefile -call prepare_sector_networks
|           (one of the commands exited with non-zero exit code; note
|   that snakemake uses bash strict mode!)
|   Shutting down, this might take some time.
|   Exiting because a job execution failed. Look above for error message
|   Complete log: /home/powerd/egon100re-run/run-pypsa-
|   eur/.snakemake/log/2024-08-14T164037.131701.snakemake.log; 4109769)

**Comment by ClaraBuettner** [comment]
Another file could not be downloaded. I added the code to download it within powerd-data and copy-pasted it in your folder to save some execution time. 
You could clear the task, or should I do that?

**Review by CarlosEpia — APPROVED** [review]
This PR was tested on the server and all the tasks ran successfully. It can be merged. There are some files that can not be downloaded when powerd-data is executed on our server, therefore they must be included in PR #262

**Pull request merged**